presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,42 +584,194 @@ private static NestedField nestedField(String name)
+            private final FunctionAndTypeManager functionAndTypeManager;
+            private final boolean isPushDownSubfieldsFromLambdasEnabled;
+
+            private SubfieldExtractor(
+                    StandardFunctionResolution functionResolution,
+                    ExpressionOptimizer expressionOptimizer,
+                    ConnectorSession connectorSession,
+                    FunctionAndTypeManager functionAndTypeManager,
+                    boolean isPushDownSubfieldsFromLambdasEnabled)
             {
                 this.functionResolution = requireNonNull(functionResolution, "functionResolution is null");
                 this.expressionOptimizer = requireNonNull(expressionOptimizer, "expressionOptimizer is null");
-                this.connectorSession = requireNonNull(connectorSession, "connectorSession is null");
+                this.connectorSession = connectorSession;
+                this.isPushDownSubfieldsFromLambdasEnabled = isPushDownSubfieldsFromLambdasEnabled;
+                this.functionAndTypeManager = functionAndTypeManager;
             }
 
             @Override
             public Void visitCall(CallExpression call, Context context)
             {
-                if (!functionResolution.isSubscriptFunction(call.getFunctionHandle())) {
-                    call.getArguments().forEach(argument -> argument.accept(this, context));
+                ComplexTypeFunctionDescriptorFunctionDescriptor functionDescriptor = functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getDescriptor();
+                if (functionResolution.isSubscriptFunction(call.getFunctionHandle()) ||
+                        functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getName().equals(ELEMENT_AT)) {
+                    Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession, functionAndTypeManager);
+                    if (isPruningSubfiedldsPossible(context)) {
+                        addRequiredSubfields(context, subfield);
+                        return null;
+                    }
+                    if (subfield.isPresent()) {
+                        context.subfields.add(subfield.get());
+                    }
+                    else {
+                        call.getArguments().forEach(argument -> argument.accept(this, context));
+                    }
                     return null;
                 }
+                if (isPushDownSubfieldsFromLambdasEnabled) {
+                    LambdaSubfieldContext lambdaSubfieldContextOriginal = context.getLambdaSubfieldContext().orElse(null);
+                    if (functionDescriptor.isAccessingInputValues()
+                            || (functionDescriptor.isTransformingInput()) && !functionDescriptor.getOutputToInputTransformationFunction().isPresent()) {
+                        context.setLambdaSubfieldContext(null);
```

Comment:
setting the context to null doesn't seem correct. first of all because we try to have things not be null. Second of all, it seems like the context should have a way to naturally indicate that subfields can't be mapped without setting it to empty.

Reply:
Changed the type to Optional.